### PR TITLE
bootstrap extension: internal listener

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
This avoid changing dispatcher interfaces when creating an internal connection at worker thread.

Additional Description:
Address the comment https://github.com/envoyproxy/envoy/pull/18105#discussion_r766098084

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
Fixes #Issue:
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by lambdai** [comment]
CC @kyessenov

**Comment by alyssawilk** [comment]
@yanavlasov ping!

File: source/extensions/io_socket/user_space/internal_listener_registry.h
Comment (kyessenov) [inline source/extensions/io_socket/user_space/internal_listener_registry.h:49]: I think we need an actual protobuf for this factory like https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/vcl/v3alpha/vcl_socket_interface.proto#extension-envoy-bootstrap-vcl. It would be good to place buffer limit there, too.

File: source/extensions/io_socket/user_space/client_connection_factory.h
Comment (yanavlasov) [inline source/extensions/io_socket/user_space/client_connection_factory.h:37]: Why do you need this? Can't you just get the slot from the InternalListenerExtension singleton?

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait-any

**Review by lambdai — COMMENTED** [review]
Thank you @yanavlasov 

I am addressing the trivial ones.

I don't have a better approach to the TLS slot. Can you share more light?

File: source/extensions/io_socket/user_space/client_connection_factory.h
Comment (lambdai) [reply source/extensions/io_socket/user_space/client_connection_factory.h:37]: Somehow my comment was lost...

This factory is constructed providing very few contexts: the dispatcher as the context is vetoed.
 
The major blocker is that Singleton object cannot be get from worker thread even though we concede that performance won't be an issue.

I am open to any alternative approach!

**Review by yanavlasov — DISMISSED** [review]

**Comment by lambdai** [comment]
Thank you @yanavlasov  !
I added a protobuf message for the bootstrap extension 
That gives up the unique occupation of `Struct` type message and allow other WIP code use `Struct` per @kyessenov

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
From API perspective, I think the naming could be improved.
Also please look at the API style [guidelines](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md#adding-an-extension-configuration-to-the-api), specifically at the updating [source/extensions/extensions_metadata.yaml](https://github.com/envoyproxy/envoy/blob/main/source/extensions/extensions_metadata.yaml) and [source/extensions/extensions_build_config.bzl](https://github.com/envoyproxy/envoy/blob/main/api/source/extensions/extensions_build_config.bzl).
/wait

File: api/envoy/extensions/io_socket/user_space/v3/bootstrap.proto
Comment (lambdai) [reply api/envoy/extensions/io_socket/user_space/v3/bootstrap.proto:19]: Thank you for the link! Working on it

Comment (lambdai) [reply api/envoy/extensions/io_socket/user_space/v3/bootstrap.proto]: Lots of renaming in https://github.com/envoyproxy/envoy/pull/19467/commits/e4894cdf5036ef7e0c50163a7bbbef181b483475

**Comment by yanavlasov** [comment]
@lambdai I think the name api/envoy/extensions/io_socket/user_space/v3/bootstrap.proto is a bit confusing. Maybe name it `internal_listener.proto`?

**Comment by lambdai** [comment]
> @lambdai I think the name api/envoy/extensions/io_socket/user_space/v3/bootstrap.proto is a bit confusing. Maybe name it `internal_listener.proto`?

Ah, yes. Thanks

**Comment by KBaichoo** [comment]
ping @yanavlasov, anything else for this PR?

**Comment by lambdai** [comment]
@adisuissa Can you release your review or we can address them in the following PRs?

File: source/extensions/io_socket/user_space/internal_listener_registry.h
Comment (adisuissa) [inline source/extensions/io_socket/user_space/internal_listener_registry.h:31]:
```suggestion
// This extension maintains the life of the ``TlsInternalListenerRegistry`` singleton.
```

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this, and apologies for the delay. I think I now understand what is happening here, and why I'm confused.

At the moment, the file `source/extensions/extensions_metadata.yaml` includes the following: https://github.com/envoyproxy/envoy/blob/2a9565da52e53b4c2ff2a263d4cc328703794bad/source/extensions/extensions_metadata.yaml#L486-L491

This implies that this `InternalListenerRegistry` type should be used for the field that has `#extension-category: envoy.io_socket` (haven't seen any field like this), and not for [bootstrap_extensions](https://github.com/envoyproxy/envoy/blob/2a9565da52e53b4c2ff2a263d4cc328703794bad/api/envoy/config/bootstrap/v3/bootstrap.proto#L282).
If you plan to use this type in the bootstrap extensions, then the category should be `envoy.bootstrap` (see WASM filter as an example).

The current comment in the `internal_listener.proto` file is somewhat confusing IMO, and I think from a user’s perspective it’s challenging to understand how this extension should be used.
I suggest splitting this up to a doc that gives an overview of the entire extension sub-system, and explains how the extension should be used.
The comments in this proto file should point to that doc, and just provide explanation about the type that is presented here (`InternalListenerRegistry`). IIUC the comment here should say that this extension should added as a bootstrap extension, if future use of `EnvoyInternalAddress` is needed.

**Comment by lambdai** [comment]
Thanks! 
Summary:
So the conclusion is to create new bootstrap extension. The new extension will depend on the existing extension `envoy.io_socket.user_space`.
The new bootstrap extension should include the instruction that it requires `envoy.io_socket.user_space`

**Review by adisuissa — COMMENTED** [review]
API-wise LGTM.
@htuch this introduces a registry that can be added as a [bootstrap-extension](https://github.com/envoyproxy/envoy/blob/b8c4d4bae0c48c2a97126e486651987ceab7b493/api/envoy/config/bootstrap/v3/bootstrap.proto#L282). The package location of the new proto (`envoy.extensions.bootstrap.internal_listener_registry.v3`) LGTM, but can you double check?

**Comment by htuch** [comment]
@adisuissa yes, I think this is the right location. OTOH, when Iook around at other bootstrap extensions, I see https://github.com/envoyproxy/envoy/blob/9d5627a0879b0a029e90515137c108e1d2884bfc/api/envoy/extensions/wasm/v3/wasm.proto#L19. This seems to be misplaced - the actual extension proto (as opposed to supporting generic Wasm protos) should also be in `envoy/extensions/bootstrap`, CC @PiotrSikora. I don't think it can be moved, but maybe we should add a comment explaining that this is an organizational exception.

File: docs/root/configuration/other_features/internal_listener.rst
Comment (htuch) [inline docs/root/configuration/other_features/internal_listener.rst:21]: Can you inline or reference an example config that demonstrates a full example?

File: source/extensions/bootstrap/internal_listener_registry/BUILD
Comment (htuch) [inline source/extensions/bootstrap/internal_listener_registry/BUILD:38]: Is this a bootstrap extension?

**Review by htuch — COMMENTED** [review]
I'm a bit  confused about the client connection factory. Is this intended to be a new extension point? It seems to be bundled with the listener registry, so is never explicitly configured as an extension?

File: docs/root/configuration/other_features/internal_listener.rst
Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:21]: add an example envoy static config yaml in the below section.

**Comment by lambdai** [comment]
> I'm a bit confused about the client connection factory. Is this intended to be a new extension point? It seems to be bundled with the listener registry, so is never explicitly configured as an extension?

Yes, this client factory is coupled with this listener registry. 
This PR gives the user flexibility to opt-out the factory and registry bundle.

This connection factory is not functional if this listener registry is not provided, or vice versa.

We can postpone the split once there is substitute of either the factory or listener registry. WDTY?

File: docs/root/configuration/other_features/internal_listener.rst
Comment (htuch) [inline docs/root/configuration/other_features/internal_listener.rst:23]: Best to switch to verifiable config fragments, @phlax what is the current best practice for this?

File: docs/root/configuration/other_features/internal_listener.rst
Comment (htuch) [inline docs/root/configuration/other_features/internal_listener.rst:16]: @lambdai what confuses me is that you explain here there are separate components, but there is in fact only a single extension.

The bundling is a bit strange, in that we are mixing upstream and downstream concerns, we typically have had independent registration in situations like this before. On the downside, splitting causes potential for misconfiguration. OTOH it's kind of confusing to have something that is nominally a "bootstrap extension" providing an "internal listener registry" also magically creating a new client connection factory type. This is really just applying principle-of-least-surprise.

Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:16]: This client connection factory is useless without this bootstrap extension because there is nowhere to lookup the internal listeners.

The very initial version of this PR is https://github.com/envoyproxy/envoy/pull/18105, which doesn't put this extension functionality into the dispatcher interface. [It was considered as a pollution to the Envoy core.](https://github.com/envoyproxy/envoy/pull/18105#discussion_r766098084) 

Now the bootstrap extension is added to allow Envoy user to disable this functionality of internal listener registration.

The client connection factory is actually the interface to the internal listener. Without it, the extension functionality is not accessible. 

Do you think we should rename this extension bootstrap.internal_client_connection_factory and doesn't change the implementation that provides the per worker thread registry?

Comment (htuch) [reply docs/root/configuration/other_features/internal_listener.rst:16]: Maybe some of my reticence can be explained by naming, I agree. It would be fine to have a single `bootstrap.internal_listener` extension that registers both. I just think it strange a "registry" is somehow also magically inserting a client connection factory.

Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:16]: Alternatively, what do you think we should do?

Split the two and throw exception when any of them cannot find each other? The client factory doesn't have the luxury to reject LDS/CDS config. 

Ideally we can reject EDS or CDS config that containing internal address, but there is also opinion that rejecting after  valiate each endpoint is expensive. Also the internal address could be referred from any config so it's not realistic to reject a http filter config that contains an internal address at configuration time.

Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:16]: re: renaming

make sense to me. Let me update

File: docs/root/configuration/other_features/internal_listener.rst
Comment (htuch) [inline docs/root/configuration/other_features/internal_listener.rst:17]:
```suggestion
internal listener address. This client connection factory is installed automatically when **envoy.bootstrap.internal_listerner** is specified.
```

File: docs/root/configuration/other_features/internal_listener.rst
Comment (htuch) [reply docs/root/configuration/other_features/internal_listener.rst:23]: Please address this one and then API LGTM>

File: docs/root/configuration/other_features/internal_listener.rst
Comment (htuch) [inline docs/root/configuration/other_features/internal_listener.rst:92]: s/http/HTTP/

File: docs/root/configuration/other_features/internal_listener.rst
Comment (htuch) [inline docs/root/configuration/other_features/internal_listener.rst:94]: Please link to https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/http_conn_man.html and also just spell out "HTTP connection manager". These are user rather than developer docs.

File: docs/root/configuration/other_features/internal_listener.rst
Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:23]: Can I start put the yaml in config/ directory and link the yaml like https://github.com/envoyproxy/envoy/blob/main/docs/root/intro/arch_overview/http/upgrades.rst#tunneling-tcp-over-http ?

And add the verifiable config as follow up?

Comment (phlax) [reply docs/root/configuration/other_features/internal_listener.rst:23]: i think best just add the config initially, but ill leave it to you to decide

to do so do

```console
git grep literalinclude docs/
```
you should find a few existing examples

File: docs/root/configuration/other_features/internal_listener.rst
Comment (phlax) [inline docs/root/configuration/other_features/internal_listener.rst:13]: i think preferably use headers for headers rather than bolding

File: CODEOWNERS
Comment (phlax) [inline CODEOWNERS:170]: flagging the codeowners errors

Comment (phlax) [reply CODEOWNERS:170]: seems like this issue is independent of this pr

**Comment by phlax** [comment]
docs are rendered here https://storage.googleapis.com/envoy-pr/19467/docs/index.html

File: docs/root/configuration/other_features/internal_listener.rst
Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst]: done

Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst]: I was almost there. The yaml is magically tested. However, I am seeing a failure related to a runtime guard. It can be reproduced at local. 
This complains even if I explicitly enable it here https://github.com/lambdai/envoy-dai/blob/intlisbootstrapext/docs/root/configuration/other_features/_include/internal-listener-config.yaml#L62
```
[2022-03-10 22:15:15.856][12][critical][assert] [source/server/listener_manager_impl.cc:339] assert failure: !config.address().has_envoy_internal_address(). Details: listener singleton_internal_encap has envoy internal address server_listener_name: "singleton_internal_encap"
. This runtime feature is disabled.
```

Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst]: @phlax So the internal-listener-config.yaml can be loaded by buildout envoy but not in config_test

Is there anything I can specialize for internal-listener-config.yaml ?

Comment (phlax) [reply docs/root/configuration/other_features/internal_listener.rst]: not sure - usually the only time it doesnt work with config_test is if there are some resources - ie certs/wasm/etc - that need to be loaded - and then its a matter of ensuring that it points to available resources.

im not so familiar with how `config_test` itself works, altho my feeling is it should work similarly enough to the built binary that configs should be compatible without modification

that it doesnt suggests to me that its not the same binary being used, altho i confess i also dont know too much about runtime guarding either

Comment (phlax) [reply docs/root/configuration/other_features/internal_listener.rst]: im wondering if its possible that config_test cant handle `bootstrap_extension` ?

i can see one of the example yaml files in the docs has a `layered_runtime` but i dont see any with `bootstrap_extension` so im wondering if that is the culprit

Comment (htuch) [reply docs/root/configuration/other_features/internal_listener.rst]: I don't think there is explicit support for bootstrap extension, but there is a fair bit of mocking going on there, so it's possible some dependency needs adding or tuning.

Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst]: @htuch @phlax 
Can I revert this verified_config and address that later? IMO the efforts is far off the topic of this PR

I can explore in another dedicated one and it would be easier to review

Comment (kyessenov) [reply docs/root/configuration/other_features/internal_listener.rst]: I can help with a verified fragment later. We probably need an end-to-end sample as well. This PR is blocking implementation for #19435.

Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst]: Updating...

Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:28]: create https://github.com/envoyproxy/envoy/issues/20335 to track the next step

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by lambdai** [comment]
@adisuissa Do you have further comment?

File: test/integration/BUILD
Comment (yanavlasov) [inline test/integration/BUILD:1685]: I think this test will fail if internal listener or userspace socket extensions are disabled. I think this test needs to be moved into an extension specific directory, such that it is automatically skipped if extension is disabled. This will also avoid the visibility change you had to make for extension.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/server/connection_handler_impl.cc
Comment (yanavlasov) [inline source/server/connection_handler_impl.cc:47]: Will this test work if internal listener extension is disabled?

File: source/server/listener_impl.cc
Comment (lambdai) [inline source/server/listener_impl.cc:495]: here @yanavlasov

File: source/server/connection_handler_impl.cc
Comment (lambdai) [reply source/server/connection_handler_impl.cc:47]: Connection Handler doesn't have the flow to fail the addListener().
This `config.internalListenerConfig().has_value()` branch should be handled by ListenerManagerImpl which would reject such config on LDS update.

File: test/integration/BUILD
Comment (lambdai) [reply test/integration/BUILD:1685]: I imagine this test binary built the new extensions, which is independent from the disabled list in //source/exe:envoy-static

Comment (lambdai) [reply test/integration/BUILD:1685]: OK, i see the integration test living in the per extension path. Let me migrate this test there

**Review by lambdai — COMMENTED** [review]
Thanks! I am moving the test case to the extension directory

File: source/server/connection_handler_impl.cc
Comment (lambdai) [reply source/server/connection_handler_impl.cc:47]: Confirmed by deleting source/extensions/bootstrap/internal_listener and the item in source/extensions/extensions_build_config.bzl

All tests passed

File: test/integration/BUILD
Comment (lambdai) [reply test/integration/BUILD]: done in https://github.com/envoyproxy/envoy/pull/19467/commits/fa940135e1ddd4ef63939b034cb4b0b41927c52a

**Comment by lambdai** [comment]
@yanavlasov I confirmed by removing the below

* source/extensions/bootstrap/internal_listener/
* test/extensions/bootstrap/internal_listener/
* `"envoy.bootstrap.internal_listener"` in source/extensions/extensions_build_config.bzl 

I believe that what we selectively import.

All tests passed on my side.

893 tests pass.

Do you have other concerns?

**Review by adisuissa — COMMENTED** [review]
Thanks for improving the doc. I think it is now easier to understand how to use this extension.
/lgtm api

File: test/integration/BUILD
Comment (lambdai) [reply test/integration/BUILD]: @yanavlasov  Moved this test in the extension directory and simulated a import without this extension. Is it good to go? Thanks

**Comment by phlax** [comment]
@yanavlasov @adisuissa looks like tests are passing, and feedback has been addressed - gtm ?

File: docs/root/configuration/other_features/internal_listener.rst
Comment (phlax) [inline docs/root/configuration/other_features/internal_listener.rst:17]: nit: s/envoy/Envoy/

File: docs/root/configuration/other_features/internal_listener.rst
Comment (phlax) [inline docs/root/configuration/other_features/internal_listener.rst:22]: i think this should be an inline literal - rather than bolded - ie ` ``envoy.bootstrap.internal_listener`` `

File: docs/root/configuration/other_features/internal_listener.rst
Comment (phlax) [inline docs/root/configuration/other_features/internal_listener.rst:26]: not sure if `tcp_proxy` should be ` ``tcp_proxy`` ` or `TCP proxy`

File: docs/root/configuration/other_features/internal_listener.rst
Comment (phlax) [reply docs/root/configuration/other_features/internal_listener.rst:22]: also noticing there is a typo in `listerner`

File: docs/root/configuration/other_features/internal_listener.rst
Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:17]: done

File: docs/root/configuration/other_features/internal_listener.rst
Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:22]: thanks! Fixec

File: docs/root/configuration/other_features/internal_listener.rst
Comment (lambdai) [reply docs/root/configuration/other_features/internal_listener.rst:26]: Seem all the rest in doc are "TCP proxy",  no inline literal. 
I update all 3 occurrences in this doc

**Review by adisuissa — APPROVED** [review]
Thanks!
API, docs and code changes LGTM.
@yanavlasov for another pass

**Comment by lambdai** [comment]
Thank you @adisuissa

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
